OpenRA.Mods.Common/Traits/Production.cs

```diff
@@ -120,25 +120,62 @@ public virtual void DoProduction(Actor self, ActorInfo producee, ExitInfo exitin
 
 				var notifyOthers = self.World.ActorsWithTrait<INotifyOtherProduction>();
 				foreach (var notify in notifyOthers)
-					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit);
+					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit, type);
 
 				foreach (var t in newUnit.TraitsImplementing<INotifyBuildComplete>())
 					t.BuildingComplete(newUnit);
 			});
 		}
 
-		public virtual bool Produce(Actor self, ActorInfo producee, string factionVariant)
+		protected virtual ExitInfo SelectExit(Actor self, ActorInfo producee, string type, Func<ExitInfo, bool> p)
+		{
+			if (type != null)
+			{
+				var exit = self.RandomExitOrDefault(type, p);
+				if (exit != null)
+					return exit;
+			}
+
+			var bi = producee.TraitInfoOrDefault<BuildableInfo>();
+			if (bi != null)
+			{
+				if (Info.Produces.Contains(bi.BuildAtProductionType))
```

Comment:
Do we need to move the fundamental production type down to the Production trait then, and then allow the same ProductionQueue to share multiple types that then use different exits (with the kennel not having any Infantry exits, and the barracks not having any Dog exits)?